magenta/models/music_vae/music_vae_export.py

```diff
@@ -0,0 +1,220 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import json
+import tensorflow as tf
+from tensorflow.python import saved_model
+from magenta.models.music_vae import configs
+
+
+# Function builds the VAE graph from a configuration object for exporting as a saved model
+def build_vae_graph(vae_config, batch_size, **sample_kwargs):
+    vae_graph = tf.Graph()
+    with vae_graph.as_default():
+        model = vae_config.model
+        model.build(
+            vae_config.hparams,
+            vae_config.data_converter.output_depth,
+            is_training=False)
+
+        # Define nodes for the model inputs
+        temperature = tf.placeholder(tf.float32, shape=())
+
+        if vae_config.hparams.z_size:
+            z_input = tf.placeholder(tf.float32, shape=[batch_size, vae_config.hparams.z_size])
+        else:
+            z_input = None
+
+        if vae_config.data_converter.control_depth > 0:
+            c_input = tf.placeholder(tf.float32, shape=[None, vae_config.data_converter.control_depth])
+        else:
+            c_input = None
+
+        inputs = tf.placeholder(
+            tf.float32,
+            shape=[batch_size, None, vae_config.data_converter.input_depth])
+        controls = tf.placeholder(
+            tf.float32,
+            shape=[batch_size, None, vae_config.data_converter.control_depth])
+        inputs_length = tf.placeholder(
+            tf.int32,
+            shape=[batch_size] + list(vae_config.data_converter.length_shape))
+
+        max_length = tf.placeholder(tf.int32, shape=())
+
+        # used in the trained_model decode functions, run in the stored session
+        outputs, decoder_results = model.sample(
+            batch_size,
+            max_length=max_length,
+            z=z_input,
+            c_input=c_input,
+            temperature=temperature,
+            **sample_kwargs)
+
+        if vae_config.hparams.z_size:
+            q_z = model.encode(inputs, inputs_length, controls)
+            mu = q_z.loc
+            sigma = q_z.scale.diag
+            z = q_z.sample()
+
+        signature_defs = build_signature_defs(
+            inputs,
+            controls,
+            inputs_length,
+            c_input,
+            z_input,
+            temperature,
+            max_length,
+            outputs,
+            decoder_results,
+            mu,
+            sigma,
+            z)
+
+    return vae_graph, signature_defs
+
+
+# This function has too many arguments, pass a data structure later
```

Comment:
We will need to fix this before submitting to Google. I'm assuming Magenta is not py3.7 compatible (if it is, just use a @dataclass!) so let's use a namedtuple I guess?